File: Group Anagrams - Leetcode 49/Group Anagrams - Leetcode 49.py
# Brute Force Solution
class Solution:
    def groupAnagrams(self, strs: List[str]) -> List[List[str]]:
        ans = defaultdict(list)
        for s in strs:
            key = ''.join(sorted(s))
            ans[key].append(s)

        return list(ans.values())
# Time: O(n * (m log m))
# Space: O(n * m)
# n is the number of strings, m is the length of largest string

# Optimal Solution
from collections import defaultdict
class Solution:
    def groupAnagrams(self, strs: List[str]) -> List[List[str]]:
        anagrams_dict = defaultdict(list)
        for s in strs: # n
            count = [0] * 26
            for c in s:
                count[ord(c) - ord("a")] += 1
            key = tuple(count)
            anagrams_dict[key].append(s)
        # we need to convert the dict.values iterable to list, otherwise we get the following error
        # error : "dict_values([['eat', 'tea', 'ate'], ['tan', 'nat'], ['bat']]) is not valid value for the expected return type list<list<string>>"
        return list(anagrams_dict.values())
# n is the number of strings, m is the length of largest string
# Time Complexity: O(n * m)
# Space Complexity: O(n * m)

File: Group Anagrams - Leetcode 49/Group Anagrams - Leetcode 49.js
/**
 * @param {string[]} strs
 * @return {string[][]}
 */
var groupAnagrams = function(strs) {
    const anagramsMap = new Map();
    
    strs.forEach(s => {
        const count = new Array(26).fill(0);
        for (let c of s) {
            count[c.charCodeAt(0) - 'a'.charCodeAt(0)]++;
        }
        const key = count.join(',');
        if (!anagramsMap.has(key)) {
            anagramsMap.set(key, []);
        }
        anagramsMap.get(key).push(s);
    });
    
    return Array.from(anagramsMap.values());
};
